Mathlib/Analysis/Convex/ConvexFunction.lean

```diff
@@ -0,0 +1,541 @@
+
+Both results are stated in a gradient version and a fderiv version.
+-/
+
+variable {E : Type*} [NormedAddCommGroup E] [InnerProductSpace ℝ E] [CompleteSpace E]
+
+section FirstOrderCondition
+
+open InnerProductSpace
+
+variable {f : E → ℝ} {f' : E → (E →L[ℝ] ℝ)} {x y x': E} {s : Set E}
+
+private lemma point_proportion {a b: ℝ} (_ : 0 ≤ a) (_ : 0 ≤ b) (sumab: a + b = 1)
+  (hpoint : x' = a • x + b • y) :  x - x' =  b • (x - y) := by
+    calc
+      x - x' = x - (a • x + b • y):= by rw [hpoint]
+      _ = x - a • x - b • y:= sub_add_eq_sub_sub x (a • x) (b • y)
+      _ = (1 : ℝ) • x - a • x - b • y:= by rw[one_smul]
+      _ = (1 - a) • x - b • y:= by rw[sub_smul 1 a]
+      _ = b • x - b • y:= by rw [← sumab]; ring_nf
+      _ = b • (x - y):= Eq.symm (smul_sub b x y)
+
+theorem first_order_condition (h : HasFDerivAt f (f' x) x) (hf : ConvexOn ℝ s f)
+    (xs : x ∈ s) : ∀ (y : E), y ∈ s → f x + f' x (y - x) ≤ f y := by
+  have h₁ : ∀ ε > (0 : ℝ), ∃ δ > (0 : ℝ), ∀ (x' : E), ‖x - x'‖ ≤ δ
+       → ‖f x' - f x - (f' x) (x' - x)‖ ≤ ε * ‖x - x'‖:= by
+    rw [HasFDerivAt, hasFDerivAtFilter_iff_isLittleO, Asymptotics.isLittleO_iff] at h
+    intro ε epos
+    specialize h epos
+    rw [Filter.Eventually] at h
+    let t := {x_1 | ‖f x_1 - f x - (f' x) (x_1 - x)‖ ≤ ε * ‖x_1 - x‖}
+    have h₁: ∃ ε1 > (0 : ℝ), Metric.ball x ε1 ⊆ t := Iff.mp Metric.mem_nhds_iff h
+    rcases h₁ with ⟨e1, e1pos, h₁⟩
+    use (e1 / 2); constructor
+    exact (half_pos e1pos)
+    intro x' xnhds
+    have h₂: x' ∈ Metric.ball x e1:= by
+      rw [Metric.mem_ball, dist_comm]
+      rw [← dist_eq_norm] at xnhds
+      apply lt_of_le_of_lt xnhds (half_lt_self e1pos)
+    have h₃: x' ∈ t := h₁ h₂
+    rw [Set.mem_setOf] at h₃
+    rw [norm_sub_rev x]
+    exact h₃
+  intro y ys
+  by_cases h₂: y = x
+  rw [h₂, sub_self, ContinuousLinearMap.map_zero (f' x), add_zero]
+  have h₃: 0 < ‖x - y‖:= by
+    rw [norm_sub_pos_iff,Ne]
+    exact Iff.mpr ne_comm h₂
+  by_contra H
+  push_neg at H
+  rw [ConvexOn] at hf
+  rcases hf with ⟨ _, cxf⟩
+  specialize cxf xs ys
+  let ε := f x + (f' x) (y - x) - f y
+  have epos : 0 < ε := by
+    exact Iff.mpr sub_pos H
+  have lnmp: ∀ c : ℝ , f' x (c • (y - x)) = c * (f' x (y - x)):= by
+    intro c
+    rw [map_smul]
+    rfl
+  let e1:= ε / (2 * ‖x - y‖)
+  have npos: 0 < 2 * ‖x - y‖ := mul_pos two_pos h₃
+  have e1pos: 0 < e1:= div_pos epos npos
+  specialize h₁ e1 e1pos
+  rcases h₁ with ⟨δ , dpos, converge⟩
+  let b1 := δ / ‖x - y‖
+  have b1pos: 0 < b1 := div_pos dpos h₃
+  let b := min b1 (1 : ℝ)
+  let a := 1 - b
+  have sum_a_b : a + b = 1:= sub_add_cancel 1 b
+  have b_nonneg: 0 ≤ b := le_min (LT.lt.le b1pos) zero_le_one
+  have a_nonneg : 0 ≤ a := by
+    have h1: 0 + b ≤ a + b := by
+      rw[zero_add, sum_a_b]
+      exact min_le_right b1 1
+    rw[add_le_add_iff_right b] at h1
+    exact h1
+  specialize cxf a_nonneg b_nonneg sum_a_b
+  let x' := a • x + b • y
+  have x'rfl : x' = a • x + b • y := rfl
```

Comment:
There's a nice tactic for precisely this. I would call the equality statement `hx'` or something starting with `h` instead of `x'rfl`.
```suggestion
  set x' := a • x + b • y with x'rfl
```